Add YOLOS

NielsRogge opened the pull request:
# What does this PR do?

This PR adds [YOLOS](https://github.com/hustvl/YOLOS), an awesome and simple object detector.

YOLOS is just a single Transformer encoder (ViT), trained using DETR's objective.

For now, I've used "vit" as `base_model_prefix`, in order to easily load weights from ViT and ViTMAE checkpoints on the hub.

Commit:
First draft

Commit:
Add YolosForObjectDetection

Commit:
Make forward pass work

Commit:
Add mid position embeddings

Commit:
Add interpolation of position encodings

Commit:
Add expected values

Commit:
Add YOLOS to tests

Commit:
Add integration test

Commit:
Support tiny model as well

Commit:
Support all models in conversion script

Commit:
Remove mid_pe_size attribute

Commit:
Make more tests pass

Commit:
Add model to README and fix config

Commit:
Add copied from statements

Commit:
Rename base_model_prefix to vit

NielsRogge commented on src/transformers/models/yolos/configuration_yolos.py:
```diff
@@ -0,0 +1,153 @@
+
+logger = logging.get_logger(__name__)
+
+VIT_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "google/vit-base-patch16-224": "https://huggingface.co/vit-base-patch16-224/resolve/main/config.json",
+    # See all YOLOS models at https://huggingface.co/models?filter=vit
+}
```
To be updated.

NielsRogge commented on tests/yolos/test_modeling_yolos.py:
```diff
@@ -0,0 +1,386 @@
+
+    @slow
+    def test_inference_object_detection_head(self):
+        # TODO rename nielsr to organization
+        model = YolosForObjectDetection.from_pretrained("nielsr/yolos-s").to(torch_device)
```
To be updated

NielsRogge commented on tests/yolos/test_modeling_yolos.py:
```diff
@@ -0,0 +1,386 @@
+class YolosModelIntegrationTest(unittest.TestCase):
+    @cached_property
+    def default_feature_extractor(self):
+        # TODO rename nielsr to organization
+        return AutoFeatureExtractor.from_pretrained("nielsr/yolos-s") if is_vision_available() else None
```
To be updated

Commit:
Add missing YOLOS_PRETRAINED_CONFIG_ARCHIVE_MAP

sgugger commented on src/transformers/models/yolos/modeling_yolos.py:
```diff
@@ -0,0 +1,1159 @@
+        )
+
+
+# copied from transformers.models.detr.modeling_detr.DetrLoss with Detr->Yolos
```
```suggestion
# Copied from transformers.models.detr.modeling_detr.DetrLoss with Detr->Yolos
```

sgugger commented on src/transformers/models/yolos/modeling_yolos.py:
```diff
@@ -0,0 +1,1159 @@
+    """
+
+    def __init__(self, matcher, num_classes, eos_coef, losses):
+        """
+        Parameters:
+        Create the criterion. A note on the num_classes parameter (copied from original repo in detr.py): "the naming
+        of the `num_classes` parameter of the criterion is somewhat misleading. it indeed corresponds to `max_obj_id +
+        1`, where max_obj_id is the maximum id for a class in your dataset. For example, COCO has a max_obj_id of 90,
+        so we pass `num_classes` to be 91. As another example, for a dataset that has a single class with id 1, you
+        should pass `num_classes` to be 2 (max_obj_id + 1). For more details on this, check the following discussion
+        https://github.com/facebookresearch/detr/issues/108#issuecomment-650269223"
+            matcher: module able to compute a matching between targets and proposals. num_classes: number of object
+            categories, omitting the special no-object category. weight_dict: dict containing as key the names of the
+            losses and as values their relative weight. eos_coef: relative classification weight applied to the
+            no-object category. losses: list of all the losses to be applied. See get_loss for list of available
+            losses.
+        """
```
This should be in the main docstring above, and it should be better formatted.

sgugger commented on src/transformers/models/yolos/modeling_yolos.py:
```diff
@@ -0,0 +1,1159 @@
+
+    def forward(self, outputs, targets):
+        """
+        Parameters:
+        This performs the loss computation.
+             outputs: dict of tensors, see the output specification of the model for the format targets: list of dicts,
+             such that len(targets) == batch_size.
+                      The expected keys in each dict depends on the losses applied, see each loss' doc
```
THis is also badly formatted. The introduction goes before `Parameters`, and then the docstring should follow our usual syntax.

sgugger commented on src/transformers/models/yolos/modeling_yolos.py:
```diff
@@ -0,0 +1,1159 @@
+        return losses
+
+
+# copied from transformers.models.detr.modeling_detr.DetrMLPPredictionHead with Detr->Yolos
```
```suggestion
# Copied from transformers.models.detr.modeling_detr.DetrMLPPredictionHead with Detr->Yolos
```

sgugger commented on src/transformers/models/yolos/modeling_yolos.py:
```diff
@@ -0,0 +1,1159 @@
+        return x
+
+
+# copied from transformers.models.detr.modeling_detr.DetrHungarianMatcher with Detr->Yolos
```
```suggestion
# Copied from transformers.models.detr.modeling_detr.DetrHungarianMatcher with Detr->Yolos
```

sgugger commented on src/transformers/models/yolos/modeling_yolos.py:
```diff
@@ -0,0 +1,1159 @@
+    @torch.no_grad()
+    def forward(self, outputs, targets):
+        """
+        Params:
+        Performs the matching.
+            outputs: This is a dict that contains at least these entries:
+                 "logits": Tensor of dim [batch_size, num_queries, num_classes] with the classification logits
+                 "pred_boxes": Tensor of dim [batch_size, num_queries, 4] with the predicted box coordinates
+            targets: This is a list of targets (len(targets) = batch_size), where each target is a dict containing:
+                 "class_labels": Tensor of dim [num_target_boxes] (where num_target_boxes is the number of ground-truth
+                 objects in the target) containing the class labels "boxes": Tensor of dim [num_target_boxes, 4]
+                 containing the target box coordinates
+        Returns:
+            A list of size batch_size, containing tuples of (index_i, index_j) where:
+                - index_i is the indices of the selected predictions (in order)
+                - index_j is the indices of the corresponding selected targets (in order)
+            For each batch element, it holds: len(index_i) = len(index_j) = min(num_queries, num_target_boxes)
```
Same as above.

sgugger commented on src/transformers/models/yolos/modeling_yolos.py:
```diff
@@ -0,0 +1,1159 @@
+        return t if t.dtype in (torch.int32, torch.int64) else t.int()
+
+
+# copied from transformers.models.detr.modeling_detr.box_area
```
```suggestion
# Copied from transformers.models.detr.modeling_detr.box_area
```

sgugger commented on src/transformers/models/yolos/modeling_yolos.py:
```diff
@@ -0,0 +1,1159 @@
+    return (boxes[:, 2] - boxes[:, 0]) * (boxes[:, 3] - boxes[:, 1])
+
+
+# copied from transformers.models.detr.modeling_detr.box_iou
```
```suggestion
# Copied from transformers.models.detr.modeling_detr.box_iou
```

sgugger commented on src/transformers/models/yolos/modeling_yolos.py:
```diff
@@ -0,0 +1,1159 @@
+# copied from transformers.models.detr.modeling_detr.generalized_box_iou
+def generalized_box_iou(boxes1, boxes2):
+    """
+    Generalized IoU from https://giou.stanford.edu/. The boxes should be in [x0, y0, x1, y1] (corner) format. Returns:
+        a [N, M] pairwise matrix, where N = len(boxes1) and M = len(boxes2)
```
Badly formatted.

sgugger commented on tests/yolos/test_modeling_yolos.py:
```diff
@@ -0,0 +1,386 @@
+
+    def prepare_config_and_inputs_for_common(self):
+        config_and_inputs = self.prepare_config_and_inputs()
+        (
+            config,
+            pixel_values,
+            labels,
+        ) = config_and_inputs
```
This all fits in one line.

sgugger commented on tests/yolos/test_modeling_yolos.py:
```diff
@@ -0,0 +1,386 @@
+    attention_mask and seq_length.
+    """
+
+    all_model_classes = (
+        (
+            YolosModel,
+            YolosForObjectDetection,
+        )
+        if is_torch_available()
+        else ()
+    )
```
This all fits in one line.

sgugger reviewed (approved):
Thanks for adding this new model! There are a few badly named variables left and some docstrings to fix, but overall it's in great shape!

Commit:
Apply suggestions from code review

NielsRogge commented on src/transformers/models/yolos/modeling_yolos.py:
```diff
@@ -0,0 +1,1159 @@
+
+    def _prune_heads(self, heads_to_prune: Dict[int, List[int]]) -> None:
+        """
+        Prunes heads of the model. heads_to_prune: dict of {layer_num: list of heads to prune in this layer} See base
+        class PreTrainedModel
```
This is also how it's defined in BERT, ViT and lots of other models. To be addressed in a separate PR?

sgugger replied:
We should start here for this one and then address the others in another PR.
The problem with other PRs is that very often everyone forget to actually do them ;-)

Commit:
Apply more suggestions from code review

NielsRogge commented:
Addressed most comments. The remaining comments are about badly formatted docstrings, however these are all copied from DETR (so I can't change them due to #Copied from statements). Is it ok if I address these docstrings in a separate PR for both models?

Also pinging @Narsil as the pipeline test for YOLOS is failing. This is because YOLOS doesn't take `pixel_mask` as input, whereas DETR does. This makes YOLOS fail for the object detection pipeline.

Commit:
Convert remaining checkpoints

sgugger commented:
I'd advocate to make the changes in docstrings in DETR to be propagated to YOLOS in this PR, just to make sure we don't forget.

Narsil commented:
> Also pinging @Narsil as the pipeline test for YOLOS is failing. This is because YOLOS doesn't take pixel_mask as input, whereas DETR does. This makes YOLOS fail for the object detection pipeline.

Then the feature_extractor should not output them. The image pipeline are pretty simple and roughly simply do

`model(**feature_extractor(image))` so if the feature extractor only outputs what's needed then it should work.
That or `pixel_mask` should be handled (doesn't seem to be making sense for this model reading your comment).

NielsRogge commented:
> Then the feature_extractor should not output them.

Yeah the problem is, YOLOS uses the same feature extractor as DETR, which outputs both `pixel_values` and `pixel_mask`. Hence, I've just added ("yolos", "DetrFeatureExtractor") to the Auto Feature Extractor API. 

I think the easiest here is to add `pixel_mask=None` to the forward of YOLOS, in order to make `model(**feature_extractor(image))` work.

Commit:
Improve docstrings

Commit:
Add YolosFeatureExtractor

Commit:
Add feature extractor to docs

Commit:
Add corresponding tests

Commit:
Fix style

NielsRogge commented:
Ok so I created a new YolosFeatureExtractor, however the pipeline test is still failing:

```
def _call_impl(self, *input, **kwargs):
        forward_call = (self._slow_forward if torch._C._get_tracing_state() else self.forward)
        # If we don't have any hooks, we want to skip the rest of the logic in
        # this function, and just call forward.
        if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
                or _global_forward_hooks or _global_forward_pre_hooks):
>           return forward_call(*input, **kwargs)
E           TypeError: forward() got an unexpected keyword argument 'pixel_mask'
```
@Narsil could you help me debug this?

Narsil commented:
> @Narsil could you help me debug this? It's weird cause YolosFeatureExtractor doesn't create a pixel mask. Also, I added doc tests which are passing.

I have checked and the reason is that the tested Feature extractor is actually a detr one, not a Yolo one:

https://github.com/huggingface/transformers/pull/16848/files#diff-fcbe32a3a065f97b00f1c242ecd45858b8d5680a2437b65af183eb0c439e2be9R347

The pipeline tests rely on ModelTester to create the base objects

Commit:
Fix docs

sgugger commented on README.md:
```diff
@@ -340,7 +340,8 @@ AWARE PRE-TRAINING](https://arxiv.org/abs/2110.05752) by Sanyuan Chen, Yu Wu, Ch
 1. **[XLS-R](https://huggingface.co/docs/transformers/model_doc/xls_r)** (from Facebook AI) released with the paper [XLS-R: Self-supervised Cross-lingual Speech Representation Learning at Scale](https://arxiv.org/abs/2111.09296) by Arun Babu, Changhan Wang, Andros Tjandra, Kushal Lakhotia, Qiantong Xu, Naman Goyal, Kritika Singh, Patrick von Platen, Yatharth Saraf, Juan Pino, Alexei Baevski, Alex ...
-1. **[YOSO](https://huggingface.co/docs/transformers/model_doc/yoso)** (from the University of Wisconsin - Madison) released with the paper [You Only Sample (Almost) Once: Linear Cost Self-Attention Via Bernoulli Sampling](https://arxiv.org/abs/2111.09714) by Zhanpeng Zeng, Yunyang Xiong, Sathya N. Ravi, Shailesh Acharya, Glenn Fung, Vikas Singh.
+1. **[YOLOS](https://huggingface.co/docs/transformers/main/model_doc/yolos)** (from Huazhong University of Science & Technology) released with the paper [You Only Look at One Sequence: Rethinking Transformer in Vision through Object Detection](https://arxiv.org/abs/2106.00666) by Yuxin Fang, Bencheng Liao, Xinggang Wang, Jiemin Fang, Jiyang Qi, Rui Wu, Jianwei Niu, Wenyu Liu.
+1. **[YOSO](https://huggingface.co/docs/transformers/main/model_doc/yoso)** (from the University of Wisconsin - Madison) released with the paper [You Only Sample (Almost) Once: Linear Cost Self-Attention Via Bernoulli Sampling](https://arxiv.org/abs/2111.09714) by Zhanpeng Zeng, Yunyang Xiong, Sathya N. Ravi, Shailesh Acharya, Glenn Fung, Vikas Singh.
```
You're putting back the main here (probably some rebase going wrong)
```suggestion
1. **[YOSO](https://huggingface.co/docs/transformers/model_doc/yoso)** (from the University of Wisconsin - Madison) released with the paper [You Only Sample (Almost) Once: Linear Cost Self-Attention Via Bernoulli Sampling](https://arxiv.org/abs/2111.09714) by Zhanpeng Zeng, Yunyang Xiong, Sathya N. Ravi, Shailesh Acharya, Glenn Fung, Vikas Singh.
```
(then you'll need a `make fix-copies`).

NielsRogge replied:
Regarding this, I see there are currently 9 models that have `main/model_doc` in the README. Should I remove "main" for all of those?

Commit:
Apply suggestion from code review

Commit:
Fix bad rebase

Commit:
Fix some more bad rebase

Commit:
Fix missing character

Commit:
Improve docs and variable names

NielsRogge commented:
Failing test is unrelated, merging.
